ecu.1
.`'  .if n .ds La '
.`'  .if n .ds Ra '
.`'  .if t .ds La `
.`'  .if t .ds Ra '
.`'  .if n .ds Lq "
.`'  .if n .ds Rq "
.`'  .if t .ds Lq ``
.`'  .if t .ds Rq ''
.ds Lq ``
.ds Rq ''
.de Ch
\\$3\\*(Lq\\$1\\*(Rq\\$2
..
.TH ECU 1 "10/18/96"
.ds ]W Version 4.07
.SH "NAME"
ecu - serial asynchronous and telnet communications program
.SH "DESCRIPTION"
ECU (Extended Call Utility) is a research and engineering
communications program for several flavors of UNIX.
ECU provides the classic terminal communications facility of
passing keyboard data to a serial line (or a telnet TCP/IP
connection if configured) and incoming data to the computer video
display.  In addition, a rich set of interactive
commands, a procedure language, a dialing directory, a function key
mapping feature, and session logging are available.
.P 
The flexible procedure (script) language allows you
to automate many communications tasks.
In addition to augmenting
interactive tasks, by using shell scripts and ECU procedures, ECU
can perform batch-style communications sessions in an entirely
"unattended" fashion.
.P
For full information, refer to the
.I "ECU Technical Description"
and the
.I "ECU Procedure Language"
manuals in the ECU source distribution.
.P
This man page describes what you need to
get ecu started and connected,
and how to use the interactive commands.

.SH "Simple Startup - Initial Setup Menu"

ECU may be started in a number of ways through use of
command line switches, but the easiest is to enter
.nf

ecu

.fi
by itself.
In this case, the screen is cleared
and the following screen, called the
.B setup
.B screen ,
will be presented.
.nf

  .--[ ecu 4.00T ]-----------------------------------------------.
  |                                                              |
  |  Destination   ________________________________________      |
  |                                                              |
  |                                                              |
  |                                                              |
  |  tty: /dev/ttys0                                             |
  |                                                              |
  |  duplex: F  baud: 9600   parity: N (data bits 8)             |
  |                                                              |
  |                                                              |
  |  TAB:next END:proceed ^C: cmd mode ^D:phone dir ESC:quit ecu |
  `- logical phone directory entry, phone number or empty -------'

.fi
Several choices may be made by navigating the setup screen.
When you position to a field, helpful text is displayed
on the bottom line of the form.

The initial (default) values for some of the
fields may be modified by command line switches or
by the special procedure "_rc.ep".  Also, entering
a dialing directory entry name in the
'Destination' field and pressing Return will
override the defaults with the values in the
directory entry.  See the sections in the ECU
Technical Description related to dialing and to
the section in the ECU Procedure Language manual
titled "_rc.ep".

.B "Destination"

If a dialing directory has been configured,
a literal phone number need not be entered.
You may enter a "logical" name matching the name
of a directory entry.

However, on your first invocation,  you'll very likely have
no directory.  You have several options:

.nf

1.  Enter a literal telephone number, (e.g. 18005551212).
    Hyphens and open and close parentheses may also
    be entered.

2.  Press END to enter open the indicated serial port,
    beginning interactive mode to communicate
    directly with the attached DCE (modem).

3.  Enter a "period-containing" hostname for a telnet
    call (if configured for telnet; see 
.B "<hostname>"
below).

4.  Press ^D to enter the telephone directory to enter
    your first directory entry.

5.  Press ^C to enter command mode immediately (with no
    line attached). This has certain uses, but probably
    not for early use.

6.  Press ESCape to exit ecu.
.fi

.B "Other Fields"

If your call is an async serial, not telnet, call,
other fields in the setup form apply.

The 'tty' field may be used to select an outgoing line
other than the default.
For important considerations on line choice, see the
ECU Technical Description
sections titled
.I "Choosing a Dialout Line"
and
.I "DCDwatch."

The 'duplex' field may be used to select a
duplex value other than the default 'F' (full).

The 'baud' field may be used to select a
bit rate value other than the default selected at Configure time.

The 'parity' field may be used to select a
parity value other than the default selected at Configure time.

.B "Keyboard Functions"

Special keyboard characters while filling in the
startup screen are:

.nf

 Enter        terminate entry in a field (or skip to
              the next field if you do not modify it)

 ^B           back up to previous field

 Cursor Up    same as ^B (if the key is available on
              your terminal and environment).

 TAB          move to next field ... if nothing typed in
              the field, do not disturb contents

 Cursor Down  same as TAB (if the key is available on
              your terminal and environment).

 END          proceed with session, dialing remote if
              logical directory entry name or literal
              telephone number enetered

 ^D           enter phone directory

 ^C           enter command mode with no line attached

 ESC          quit ecu without starting a session
.fi

.SH "Command Line Switches and Arguments"

ECU can be started in a number of ways:
.nf

o  with no switches or arguments
       select options interactively; manual command to begin
       connection

o  with switches (excluding -p) and no arguments
       override some defaults for options but still enter
       interactive option selection; manual command to begin
       connection

o  zero or more switches (excluding -p) and one argument
       fully automatic startup to connect to the remote
       specified by the argument (a telephone number
       or dialing directory entry)

o  zero or more switches, with -p the last switch
       fully automatic startup by executing procedure whose
       name is specified by the first argument; the initial
       procedure receives the remainder of the command line
       arguments
.fi

.B "Usage Summary"

.nf

Usage: ecu [-l <ttyname>] [-b <baud_rate>] [-eon]
           [-h] [-t] [-P <phonedir>]
           [-F name] [-T <trace-level>] [-z]
           [-p <initial_proc> [-d] [-D] | <phone_number>
           <logical> | <hostname> ]
-D unconditionally stop execution when -p initial
   procedure is done
-F sets an alternate funckeymap name for the *keyboard*
-P choose alternate phone directory (<phonedir> must be
   a full pathname)
-T set procedure tracing to level: 0=none, 1=standard,
   2=ECU-debugging
-b choose bit rate (any UNIX rate 110-38400)
-d stop execution if -p initial procedure fails
-e even parity   -o odd parity   -n no parity
-h half duplex ... default is full duplex
-l choose line (/dev/<ttyname>)
-p execute an initial procedure
-t append NL to incoming and outgoing CR characters
-z if telnet connection, show options traffic 
<phone_number> is either an actual number or a dialing
   directory entry name

If configured for telnet use, if a period '.' appears in
phone number, contact host by that name using telnet; a
trailing period will be removed.
.fi

.B "Switches"

Many switches are used to override defaults specified
at the time the Confifgure procedure was used.
Note these defaults may also be overriden interactively if the command
does not specify automatic startup option (-p or dialing argument).
Also, the value chosen by a dialing directory entry or an
initial procedure will override
the Configured default and/or the value specified by this switch.

.B "-b <baud>"

This switch overrides the default bit rate.
Any valid UNIX rate may be chosen.  
ECU chooses the number of stop bits based on the bit rate.  Rates below
300 baud get 2 stop bits.  Rates 300 baud and above get one stop bit.

.B "-d"

The
.I "-d switch"
instructs ECU to "die" (terminate with error status) if
an initial procedure (-p) fails.  This switch ensures a batch
ECU execution will hang up any connection and terminate if
a procedure error occurs. See also -D below.

Absence of the
.I "-d"
and
.I "-D"
switches
causes ecu (upon any completion of the
initial procedure) to enter the 1) interactive mode if a line
was successfully attached by the procedure or 2) the setup screen
if no line was attached. 

.B "-e, -o, -n"

Normally, ECU starts up with data bits and parity chosen at
Configure time.  The 
.I -e ,
.I -o 
and
.I -n 
.I switches
allow you to override the default.

Since combinations like eight data bit and even parity or
seven data bits and no parity are not in  ECU's
capability to use, the parity selection also dictates the
choice of data bits. Even or odd parity implies seven data bits.
No parity implies eight data bits.

.B "-h"

Normally, ECU starts up in the full duplex mode.
If half duplex is desired, the
.I "-h switch"
is used.

.B "-l ttyspec"

When ECU starts up, it normally chooses a line as described
in "Choosing a Dialout Line"
in the ECU Technical Description.
Specifying the
.I -l
.I switch
overrides the default tty specified
at Configure time.  Depending upon other command line options,
this switch may be nothing more than a hint.

For important considerations on line choice, see the sections
titled "Choosing a Dialout Line" and "DCDwatch".

Two styles of argument to -l allow line selection by two methods.

Of course, no -l is meaningful for a telnet destination.

.*s 5 "by specific device"

The ususal argument to the switch is the base name
of the tty (e.g., "tty1a" or "acu0").  

Note to old users of ECU:
In previous SCO versions of ECU, since ttys were generally all
named with the prefix "tty", ECU allowed you to
omit the "tty" (e.g., "1a" or "4g"). This is no longer the case.

.*s 5 "by Devices type"

Alternatively, if your platform supports
HDB UUCP, you may choose a line by UUCP Devices type by
specifying the type with a leading equals sign.
You must also explicitly set the bit rate with -b (unless
the default bit rate is acceptable).
.nf

    ecu -l=VoiceDial -b2400

.fi
searches the UUCP Devices file for an entry whose type is
``VoiceDial'' that accepts 2400 baud.

.B "-p <proc>"

The 
.I "-p switch"
causes ECU to execute the procedure <proc> ("<proc>.ep")
immediately upon startup.  Such a procedure is termed the

CHANGES
Changes for 3.30 since 3.24:

1.  An ANSI CUP (cursor position) with x or y values exceeding
screen size should set the cursor to the bottom or right of the
screen, not the top or left.  ECU did the latter; it is fixed.
The DSR (device status report) handler was broken also.  It is
fixed, and differentiation is now made between a type 5 and 6 DSR
request.  As a result of all this, the X11 resize program will work
with ECU.

2. The command line -l handler now treats an argument
with a leading equals sign as a Devices type rather than
an explicit line.  That is,

    ecu -l=VSAT

will search for an ``idle'' line matching any entry
in your HDB Devices file whose first field is "VSAT".
The term ``idle'' is strictly correct only on SCO machines
or where the utmp/ecuungetty code has been made to work.
The baud rate is automatically set to the highest speed
specified by the Devices entry, overriding any value set
by the -b command line switch.

Note that specifying a logical (directory entry name)
rather than an actual (numeric) phone number will cause the
directory entry choice of line to override the line specified on
by -l.  This has always been true of -lttyxx treatment as well.

Perhaps one day, I'll have time to make a directory entry
specifying "Any" *not* override -l and initiate it's own search for a
free line (based as it is on baud rate and type == "ACU" alone).
There is no time now.

3. The command line usage has been updated to show what
command line variables are overriden by a logical phone number.
Again, no time to put it in the manual.

4. The integer operators << and >> were added.

5. The procedure command eeod (erase to end of display) was added.

6.  Previously, you could bind a function key to a character
sequence to be transmited over the line.  You may now bind a
function key to a procedure invocation.  If the first token in
the third field of a key definition is "proc", the rest of the
third field is taken to be a procedure invocation.

Consider the following example entry in ~/.ecu/keys:

test
    F1:VT100 F1:esc O q
    F2:doit:proc doit arg1 arg2 argn
                 ^  64 chars max   ^
                 `-----------------'

The maximum length of the procedure name, arguments and
whitespace must be less than 64 characters; any excess is
silently discarded.

Procedure doit might, for example, send a prompt to a remote to
begin an XMODEM transaction, then start an XMODEM receive to
get the file.  (This helps make up for the unsuitability of XMODEM's
C or NAK header for automatic download startup.)

7. The fkey definition file (~/.ecu/keys in the documentation)
can be located in the library directory (usually /usr/local/lib/ecu).
This has always been in the code.  Now it is documented.

8. The code for saving a phone directory lacked good error detection.
This has been fixed.

9.  -D__STDC__=0 was added to gcc flags.

10.  The default for SCO UNIX configuration is now 'v' not 'u',
making 3.2v4/ODT 2.0 the default.

11. A discussion of how to "finish" a remote C-kermit server was
added to README.Q-and-A.

12. Under SunOS, DTR was not properly controlled.  This has been fixed.

13. Configuration for use with gcc now requires you have gcc 1.40
or later.  If you say you have gcc 2.3.3 or later, you will get -O6
compilations (full global optimaztion).  If you encounter problems
with the software compile with -O6, reconfigure answering no to
the 2.3.3 question and remake.  Please let me know of any problems.

14. -Wuninitialized was added to gcc compilations.
This is only used if you answered yes to the 2.3.3 question

15. ECU works again on XENIX 286 thanks to vancleef@netcom.com
(Henry van Cleef)

    a. CFLAGS from "-M2let128" to "-M2e -i"
    b. ZMODEM code unsigned Baudrate to unsigned long

(Not a bad port given I haven't seen a 286 in 5 years!)

16.  An experimental USR 2400 dialer (dceUSR24v.c/dialgUSR24v)
has been added to gendial for attended voice line autodialer
applications.  It is written for speed over reliability, so so
the careful initialization found in other gendial dialers is
absent.  This means dialgUSR24v works best with a modem that is
dedicated to the voice autodial function.

As delivered, dialgUSR24v will prefix a '1' to a number passed to it if
the number is ten digits long and has [01] as its second digit.
(I wanted to word click with xterm and my phone directory is an
ancient flat file with no '1' prefix to be found.)
You may need to adopt, hack and improve.

DialgUSR24v, like all gendial dialers, are supplied "as-is" in
hope the code may be useful. (Okay, and also to promote the
excellent SCO solution to modem madness.)

17. Previously, a Devices-specified  SCO dialer program had to
have a full pathname specified for ECU to use it (otherwise,
ECU would treat the token as the name of an entry in Dialers).
Now, when ECU sees a simple name in a Devices entry, it checks
for an executable in the UUCP library directory before searching
Dialers.

18. PgUp and PgDn have been added to the directory menu repertorie.

PgUp   ^U    page up
PgDn   ^D    page down

19.  The new conxout interactive command was missing from
interactive help.

20.  I had a hack cpio entry in the main Makefile for my use

ecu.c
/*+-----------------------------------------------------------------------
	ecu.c - Extended Calling Unit/Call Utility/Call UNIX/whatever
	wht@wht.net

  Defined functions:
	main(argc, argv)
	xmtr()

   "Now tell me, old man, what's your name?"
   "Holloway, sir. Charles William Holloway."
   "Oh yes, the town's librarian."
   "I have the honor, sir, and have had for many years."
   "I believe all that time spent living through other men's lives,
other men's dreams, what a waste."
   "Sometimes, a man can learn more from other men's dreams than he
can from his own.  Come visit me sir, if you wish to improve your
education and I may improve yours."  -- Something Wicked This Way Comes

------------------------------------------------------------------------*/
/*+:EDITS:*/
/*:04-26-2000-11:15-wht@bob-RELEASE 4.42 */
/*:12-19-1997-21:45-wht@kepler-add -k */
/*:12-18-1997-05:06-wht@sidonia-AIX port: sbrk and getpgrp */
/*:05-10-1997-13:53-wht@gyro-strncpy not strcpy to shm->ttyname */
/*:01-24-1997-02:36-wht@yuriatin-SOURCE RELEASE 4.00 */
/*:10-24-1996-03:11-wht@yuriatin-T switch now incremental sw for proc_trace */
/*:09-11-1996-19:59-wht@yuriatin-3.48-major telnet,curses,structural overhaul */
/*:08-21-1996-16:57-wht@fep-add -C switch for show_config */
/*:08-20-1996-12:39-wht@kepler-locale/ctype fixes from ache@nagual.ru */
/*:12-12-1995-14:22-wht@kepler-notify [no line attached] on keystroke */
/*:12-06-1995-13:30-wht@n4hgf-termecu w/errno -1 consideration */
/*:12-03-1995-19:57-wht@gyro-use Setuid */
/*:11-23-1995-11:20-wht@kepler-source control 3.37 for tsx-11 */
/*:11-14-1995-10:23-wht@kepler-3.37.80-source control point: SOCKETS */
/*:11-13-1995-12:07-wht@kepler-xmtr() can live with no line open */
/*:11-12-1995-01:11-wht@gyro-add switch -n to turn off rcvr_ansi_filter */
/*:11-12-1995-00:18-wht@gyro-show_telnet_traffic now in shm */
/*:11-10-1995-18:32-wht@gyro-add switch -z for show_telnet_traffic */
/*:11-04-1995-21:03-root@wwtp1-move windows rattle for quirk earlier in run */
/*:10-14-1995-16:17-wht@kepler-call build_valid_baud_string */
/*:09-17-1995-16:28-wht@kepler-remove obsolete #if 0 code */
/*:06-14-1995-19:08-wht@n4hgf-do not use setuid if root is real user */
/*:06-12-1995-15:03-wht@n4hgf-if ecu has uucp euid, make use of it */
/*:06-12-1995-14:13-wht@n4hgf-no more root euid nice code */
/*:01-12-1995-15:19-wht@n4hgf-apply Andrew Chernov 8-bit clean+FreeBSD patch */
/*:05-04-1994-04:38-wht@n4hgf-ECU release 3.30 */
/*:08-16-1993-17:19-wht@n4hgf-aid share debug with report_initial_line() */
/*:08-07-1993-20:38-wht@n4hgf-add xmtr_wfp_debug_hack */
/*:12-20-1992-12:08-wht@n4hgf-continue -l=Devices support work */
/*:12-04-1992-20:57-wht@n4hgf-add -l=Devices support */
/*:09-10-1992-13:58-wht@n4hgf-ECU release 3.20 */
/*:08-22-1992-15:38-wht@n4hgf-ECU release 3.20 BETA */
/*:08-16-1992-03:43-wht@n4hgf-add -F funckeytype */
/*:08-16-1992-03:08-wht@n4hgf-head off another POSIX plot */
/*:04-30-1992-00:49-wht@n4hgf-remove obsolete -c */
/*:04-24-1992-21:59-wht@n4hgf-more SCO tty name normalizing */
/*:04-17-1992-18:19-wht@n4hgf-"default" keyset read if present */
/*:02-16-1992-02:39-wht@n4hgf-add -P phonedir switch for rll@sco */
/*:02-16-1992-01:41-wht@n4hgf-turn off xterm_title */
/*:08-11-1991-19:56-wht@n4hgf-soup up -l for ISC vs. SCO */
/*:08-06-1991-13:02-wht@n4hgf-jpm@logixwi fix: HZ getenv test wrong sense */
/*:07-29-1991-17:57-wht@n4hgf-add memstat */
/*:07-25-1991-12:55-wht@n4hgf-ECU release 3.10 */
/*:07-17-1991-07:04-wht@n4hgf-avoid SCO UNIX nap bug */
/*:04-27-1991-01:52-wht@n4hgf-overhaul revision numbers */
/*:03-17-1991-13:44-wht@n4hgf-nice and uid revision */
/*:01-09-1991-22:31-wht@n4hgf-ISC port */
/*:11-30-1990-19:04-wht@n4hgf-new ttyinit parameter - see TTYINIT_... */
/*:11-28-1990-15:58-wht@n4hgf-add non-ansi terminal support */
/*:08-14-1990-20:39-wht@n4hgf-ecu3.00-flush old edit history */

#ifdef __FreeBSD__
#include <locale.h>
#endif
#include "ecu.h"
#include "dvent.h"
#include "esd.h"
#include "procedure.h"
#include "ecukey.h"
#include "ecuxkey.h"
#include "termecu.h"
#include "ecu_pwd.h"

char *getenv();
void xmtr_SIGINT_handler();

extern char *makedate;		 /* temporary make date */
extern char *revstr;		 /* ecunumrev.c */
extern char *revision_modifier;	/* ecunumrev.c */
extern ESD *icmd_prompt;
extern char kbdeof;			 /* current input EOF */
extern char kbdeol2;		 /* current secondary input EOL */
extern char kbdeol;			 /* current input EOL */
extern char kbderase;		 /* current input ERASE */
extern char kbdintr;		 /* current input INTR */
extern char kbdkill;		 /* current input KILL */
extern char kbdquit;		 /* current input QUIT */
extern UINT tcap_LINES;
extern UINT tcap_COLS;
extern UINT32 colors_current;
extern int there_is_hdb_on_this_machine;
extern char phonedir_name[PHONEDIR_NAME_SIZE];

char *eculibdir = CFG_EcuLibDir;
char *dash_f_funckeytype;
char hello_str[128];		 /* msg printed upon BOJ */
char errmsg[128];
char initial_procedure[128] = "";
char *default_tty = CFG_DefaultTty;	/* configurable default tty name */
int init_proc_argc = 0;
char *init_proc_argv[MAX_PARGV];
CFG_PidType xmtr_pid = 0;
int quit_on_init_proc_fail = 0;
int quit_on_init_proc_done = 0;
int rc_ep_has_run = 0;
struct TIMEB starting_timeb;

extern int hertz;					 /* HZ from environ or sys/param.h */
extern UINT32 hzmsec;				 /* clock period in msec rounded up */

#ifdef CFG_Malloc3X
char *startbrk;
char *startsp;

#endif

char *_rc = "_rc";			 /* _rc.ep */

/*+-----------------------------------------------------------------------
	xmtr() --  copy stdin to comm line

  THE INITIAL PROCESS EXECUTES THIS PROCESS UNTIL PROGRAM TERMINATION

------------------------------------------------------------------------*/
void
xmtr()
{
	UINT xmtr_char;
	char nlchar = NL;
	char tmpc;

	xmtr_wfp_debug_hack();	 /* don't ask */
	ttymode(1);
	xmtr_signals();
	while (1)
	{

		xmtr_char = ttygetc(1);

		if (xmtr_char & 0x100)
		{
			kbd_escape(xmtr_char);
			continue;
		}

		if (shm->Liofd <= 0)
		{
			UINT32 colors_at_entry = colors_current;

			setcolor(colors_notify);
			fputs("[no line attached]", se);
			setcolor(colors_at_entry);
			ff(se, "\r\n");
			continue;
		}

		lputc(xmtr_char);
		if (!shm->Lfull_duplex)
		{					 /* echo character if asked */
			tmpc = xmtr_char;
			write(TTYERR, &tmpc, 1);
		}

		if (xmtr_char == CRET)
		{
#if defined(CFG_TelnetOption)
			if (shm->Ltelnet && !shm->Ltelnet_raw)
				lputc(0);
#endif
			if (shm->Ladd_nl_outgoing)
				lputc('\n');
			if (!shm->Lfull_duplex)
				write(TTYERR, &nlchar, 1);
		}
	}
	/* NOTREACHED */

}							 /* end of xmtr */

/*+-------------------------------------------------------------------------
	keyboard_test_only() - diagnose keyboard problems
--------------------------------------------------------------------------*/
void
keyboard_test_only()
{
	ttyinit(TTYUSE_NORMAL);
	if (tty_not_char_special)
	{
		ff(se, "Cannot test keyboard unless it is stanmdard input\n");
		exit(1);
	}
	ttymode(1);				 /* put user console in `raw' mode */
	kbd_test();				 /* in ecutty.c */
	ttymode(0);				 /* restore keyboard */
	termecu(0);

}							 /* end of keyboard_test_only */

/*+-------------------------------------------------------------------------
	main(argc,argv)

  main() program forks to create rcvr process; then main()
  becomes the xmtr process
------------------------------------------------------------------------*/
int
main(argc, argv)
int argc;
char **argv;
{
	int swchar;
	int itmp;
	UINT32 colors_save;
	struct passwd *pw;
	char *get_ttyname();
	extern char *optarg;
	extern int optind;

#ifdef CFG_Malloc3X

	/*
	 * absolutely must be first on the list
	 */
	startbrk = sbrk(0);		 /* initial break */
	startsp = (char *)&swchar;	/* initial sp */
	(void)mallopt(M_MXFAST, 256);
	(void)mallopt(M_NLBLKS, 64);
	(void)mallopt(M_GRAIN, 64);
#endif

#ifdef __FreeBSD__
	setlocale(LC_ALL, "");
#endif

	setbuf(stderr, NULL);	 /* rarely necessary */

	build_valid_baud_string();

	Ftime(&starting_timeb);	 /* get startup time */
	xmtr_signals();			 /* catch xmtr signals */
	xmtr_pid = (CFG_PidType) getpid();

	/*
	 * get this off quick, cause we'll be busy for a little while on 286
	 */
	build_revision_string();
	sprintf(hello_str, "ecu %s", revstr);
#ifdef M_I286
	ff(se, "%s\n", hello_str);
#endif

	/*
	 * initialize 'hertz'/'hzmsec' and  Nap() implimentation
	 */
	init_Nap();

	/*
	 * init uucp euid processing
	 */
	uid = getuid();
	euid = geteuid();
	if (pw = getpwnam("uucp"))
		uid_uucp = pw->pw_uid;
	endpwent();

	/*
	 * if we are not root, nor uucp (unlikely) and we are euid uucp, then
	 * use Setuid() to access uucp owned lines
	 */
	setuid_uucp = (uid) && (uid_uucp) && (uid != euid) && (euid == uid_uucp);
	if ((uid != euid) && Setuid(uid))
	{
		ff(se, "\r\n");
		perror("Setuid(uid) failed");
		ff(se, "\r\n");
		exit(-1);
	}

	/*
	 * Initialize shared memory segment. This must be done before any uses
	 * are made of the Lermio variable see ecushm.c and ecu.h.
	 */
	shm_init();

	/*
	 * ecu has its own keyboard handler ... blessing and curse since it
	 * depends on little outside help but can be tedious to configure for
	 * a new host or even instance
	 */
	keyset_init();			 /* intialize keyset */
	keyset_read("default");

	/*
	 * make ~/.ecu if necessary ... but chmod 700 the directory even if it
	 * already existed (needs to be very early before lots of init)
	 */
	make_ecu_subdir();

	get_curr_dir(curr_dir, sizeof(curr_dir));
	cd_array_init();		 /* read %cd directory list */

	hdb_init();
	var_init();				 /* initialize procedure variables */
	poutput_init();			 /* initialize procedure output */

	icmd_prompt = esdalloc(64);
	set_default_escape_prompt();

	/*
	 * init line variables
	 */
	memset(shm->Lline, 0, sizeof(shm->Lline));
	shm->Liofd = -1;		 /* no line open now */
	shm->Lbitrate = CFG_DefaultBitRate;	/* from config.c run */
	shm->Lparity = CFG_DefaultParity;	/* from config.c run */
	if (shm->Lparity == 'n')
		shm->Lparity = 0;
	shm->Ltelno[0] = 0;		 /* no telephone number for remote yet */
	shm->Llogical[0] = 0;	 /* no logical name for remote yet */
	shm->Lrname[0] = 0;		 /* no logical name for remote yet */
	shm->Ldescr[0] = 0;		 /* no description for remote yet */
	shm->Lconnected = 0;	 /* not connected */
	shm->Ladd_nl_incoming = 0;	/* dont add nl to incoming cr */
	shm->Ladd_nl_outgoing = 0;	/* dont add nl to outgoing cr */
	shm->Lfull_duplex = 1;	 /* assume full duplex */
	shm->Lmodem_already_init = 0;	/* modem has not been initialized */
	shm->Lxonxoff = IXON | IXOFF;	/* default to xon/xoff protocol */
	shm->xmtr_pid = (CFG_PidType) getpid();
	shm->xmtr_ppid = (CFG_PidType) getppid();
#if defined(CFG_GetpgrpVoidArg)
	shm->xmtr_pgrp = (CFG_PidType) getpgrp();
#else
	shm->xmtr_pgrp = (CFG_PidType) getpgrp(0);
#endif
	strncpy(shm->tty_name, get_ttyname(), sizeof(shm->tty_name));
	shm->tty_name[sizeof(shm->tty_name) - 1] = 0;
	shm->shm_revision = SHM_REV;
	shm->ttyuse = TTYUSE_NORMAL;
	shm->terminating = 0;

	while ((swchar = getopt(argc, argv, "kCDF:HNP:Tb:defhl:no:p:tz")) != -1)
	{
		switch (swchar)
		{
			case 'k':
				keyboard_test_only();
				/* NOTREACHED */
			case 'b':
				if (valid_baud(shm->Lbitrate = atoi(optarg)) == -1)
				{
					ff(se, "invalid bit rate %u\n", shm->Lbitrate);
					usage();
				}
				break;
			case 'l':
				shm->Lline[0] = 0;
				if ((*optarg != '=') && (*optarg != '/'))
					strcpy(shm->Lline, "/dev/");
				strncat(shm->Lline, optarg,
					sizeof(shm->Lline) - strlen(shm->Lline) -
					strlen(optarg) - 1);
				shm->Lline[sizeof(shm->Lline) - 1] = 0;
				break;
			case 'p':
				strncpy(initial_procedure, optarg, sizeof(initial_procedure));
				initial_procedure[sizeof(initial_procedure) - 1] = 0;
				break;
			case 'h':
				shm->Lfull_duplex = 0;
				break;
			case 'f':
				shm->Lfull_duplex = 1;
				break;
			case 'd':
				quit_on_init_proc_fail = 1;
				break;
			case 'D':
				quit_on_init_proc_done = 1;
				break;
			case 't':
				shm->Ladd_nl_incoming = 1;
				shm->Ladd_nl_outgoing = 1;
				break;
			case 'e':
				shm->Lparity = 'e';
				break;
			case 'o':
				shm->Lparity = 'o';
				break;
			case 'N':
				shm->ttyuse = TTYUSE_FORCE_SIMPLE;
				break;
			case 'T':
				proc_trace++;
				break;
#if defined(CFG_TelnetOption)
			case 'R':
				shm->Ltelnet_raw = 1;
				break;
#endif
			case 'P':
				strncpy(phonedir_name, optarg, PHONEDIR_NAME_SIZE);
				phonedir_name[PHONEDIR_NAME_SIZE - 1] = 0;
				break;
			case 'C':
				show_config();
				termecu(0);
				break;
			case 'F':
				dash_f_funckeytype = optarg;
				break;
			case 'n':
				shm->rcvr_ansi_filter = 0;
				break;
			case 'z':
				shm->show_telnet_traffic = 1;
				break;
			case '?':
				usage();
		}
	}

	/*
	 * check a few options for validity
	 */
	if (!initial_procedure[0] &&
		(quit_on_init_proc_done || quit_on_init_proc_fail))
	{
		ff(se, "no -D/-d without -p\n");
		usage();
	}

	ttyinit(shm->ttyuse);	 /* init console tty mode handler */
	ttymode(3);				 /* put user console in `raw' mode but SIGINT
							  * terms prog */

	if (tty_not_char_special)
		quit_on_init_proc_done = 1;
	else
	{

		/*
		 * rattle curses once - fixes quirk/bug I can't find
		 */
		windows_start();
		windows_end(0);
		fflush(so);
		tcap_clear_screen();
	}

#if defined(WHT2) || defined(XTERM_FRIEND)

	/*
	 * if xterm, put notice in title bar but this really should be done in
	 * _rc.ep
	 */
	xterm_title("ECU", 0);
#endif

	/*
	 * do the _rc.ep execution
	 */
	if (find_procedure(_rc))
	{
		if (do_proc(1, &_rc))
		{
			if (quit_on_init_proc_fail || quit_on_init_proc_done)
				termecu(TERMECU_INIT_PROC_ERROR);
		}
	}
	rc_ep_has_run = 1;

	/*
	 * check out line
	 */
	if (!shm->Lline[0])
	{
		if (!there_is_hdb_on_this_machine)
			strcpy(shm->Lline, default_tty);
		else
		{
			DVE *tdve;

			if (tdve = hdb_choose_Any(shm->Lbitrate))
				sprintf(shm->Lline, "/dev/%s", tdve->line);
			else
				strcpy(shm->Lline, default_tty);
		}
	}
	else if (shm->Lline[0] == '=')	/* Devices lookup */
	{
		DVE *tdve = 0;
		char acutype[128];

		if (!there_is_hdb_on_this_machine)
		{
			ff(se, "\r\n\n");
			tcap_stand_out();
			fputs("[cannot use -l=Devices ... HDB not available]", se);
			tcap_stand_end();
			ff(se, "\r\n");
			usage();
		}
		strncpy(acutype, shm->Lline + 1, sizeof(acutype));
		acutype[sizeof(acutype) - 1] = 0;
		if (!(tdve = getdvtype(shm->Lline + 1)))
		{
			ff(se, "\r\n\n");
			tcap_stand_out();
			fprintf(se, "[no Devices line of type '%s']", acutype);
			tcap_stand_end();
			ff(se, "\r\n");
			errno = -1;
			termecu(TERMECU_SVC_NOT_AVAIL);
		}
		shm->Lline[0] = 0;
		shm->Lbitrate = tdve->high_baud;
		setdvent();			 /* rewind for assured passthru/enddvent by
							  * hdb_choose */
		if (!(tdve = hdb_choose_Device(acutype, shm->Lbitrate)))
		{
			ff(se, "\r\n\n");
			tcap_stand_out();
			fprintf(se, "[no idle Devices line of type '%s' at %u baud]",
				acutype, shm->Lbitrate);
			tcap_stand_end();
			ff(se, "\r\n");
			errno = -1;
			termecu(TERMECU_SVC_NOT_AVAIL);
		}
		strcpy(shm->Lline, "/dev/");
		strncat(shm->Lline, tdve->line, sizeof(shm->Lline) - strlen(shm->Lline));
		shm->Lline[sizeof(shm->Lline) - 1] = 0;
	}

	report_initial_line();	 /* in #ifdef'ed debug event logging context */

	/*
	 * either present startup screen or run initial procedure or both
	 */
	if (initial_procedure[0])
	{
		init_proc_argv[0] = initial_procedure;
		init_proc_argc = 1;
		for (itmp = optind; itmp < argc; itmp++)
		{
			if (*argv[itmp] != '-')
			{
				if (init_proc_argc == MAX_PARGV)
				{
					ff(se, "too many arguments to initial procedure\r\n");
					errno = -1;
					termecu(TERMECU_USAGE);
				}
				init_proc_argv[init_proc_argc++] = argv[itmp];
			}
		}

		if (do_proc(init_proc_argc, init_proc_argv))
		{
			if (quit_on_init_proc_fail || quit_on_init_proc_done)
				termecu(TERMECU_INIT_PROC_ERROR);
		}
		proc_file_reset();
		colors_save = colors_current;
		setcolor(colors_notify);
		pputs("[procedure finished]");
		setcolor(colors_save);
		pputs("\n");
		if (quit_on_init_proc_done)
		{
			errno = -1;
			termecu(TERMECU_OK);
		}
		if (shm->Liofd < 0)
		{
			ff(se, "\r\n\n");
			tcap_stand_out();
			fputs("[no line attached by initial procedure]", se);
			tcap_stand_end();
			ff(se, "\r\n");
			if (quit_on_init_proc_fail)
				termecu(TERMECU_INIT_PROC_ERROR);
			tcap_stand_out();
			fputs("[press ESC to exit or SPACE for setup menu]", se);
			tcap_stand_end();
			itmp = ttygetc(0);
			ff(se, "\r\n");
			if (itmp == ESC)
				termecu(TERMECU_OK);
			shm->Llogical[0] = 0;
			shm->Ltelno[0] = 0;
			shm->Ldescr[0] = 0;
			setup_screen((char *)0);
		}
		else
			start_rcvr_process(0);
	}
	else
		/* no initial procedure */
	{
		if (!tty_not_char_special)
			setup_screen((optind < argc) ? argv[optind] : (char *)0);
		else
		{
			errno = -1;
			pprintf("Cannot run ecu from non-tty without initial procedure.\n");
			termecu(TERMECU_UNRECOVERABLE);
		}
	}

	/* enter xmtr operation */
	xmtr();
	termecu(TERMECU_OK);
	return (0);				 /* never get here, but keep gcc optim from
							  * complaining */
	/* NOTREACHED */
}							 /* end of main */

/* end of ecu.c */
/* vi: set tabstop=4 shiftwidth=4: */
